src/main/java/com/iota/iri/Snapshot.java

```diff
@@ -86,76 +87,141 @@
+     * Caller has to make sure that he holds a lock.
+     */
+    public void markApproved(Hash h) {
+        approvedHashes.add(h);
+    }
+
+    public Optional<Long> getStateOf(Hash h) {
+        return Optional.ofNullable(this.computedState.computeIfAbsent(h, (someValue) -> getStateOfNaive(someValue).orElse(null)));
+    }
+
+    // This optimisation means that we only add a hash to the computed state on the snapshots where it's actually
+    // used
+    private Optional<Long> getStateOfNaive(Hash h) {
+        Long l = this.computedState.get(h);
+
+        if (l != null) {
+            return Optional.of(l);
+        }
+
+        l = this.state.get(h);
+        Snapshot ref = referencedSnapshot;
+
+        while (ref != null) {
+            if (ref.computedState.containsKey(h)) {
+                Long v = ref.computedState.get(h);
+                if (l != null) {
+                    return Optional.of(l + v);
+                } else {
+                    return Optional.of(v);
+                }
+            }
+
+            Long v = ref.state.get(h);
+            if (l == null) {
+                l = v;
+            } else if (v != null) {
+                l += v;
+            }
+
+            ref = ref.referencedSnapshot;
+        }
+
+        return Optional.ofNullable(l);
     }
 
     public Snapshot patch(Map<Hash, Long> diff, int index) {
-        Map<Hash, Long> patchedState = state.entrySet().parallelStream()
-                .map( hashLongEntry ->
-                        new HashMap.SimpleEntry<>(hashLongEntry.getKey(),
-                                hashLongEntry.getValue() +
-                                        (diff.containsKey(hashLongEntry.getKey()) ?
-                                         diff.get(hashLongEntry.getKey()) : 0)) )
+        /*Map<Hash, Long> patchedState = diff.entrySet().parallelStream()
+                .map(hashLongEntry -> new HashMap.SimpleEntry<>(hashLongEntry.getKey(),
+                        hashLongEntry.getValue() + getStateOf(hashLongEntry.getKey()).orElse(0L)))
                 .filter(e -> e.getValue() != 0L)
-                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
-        diff.entrySet().stream()
-                .filter(e -> e.getValue() > 0L)
-                .forEach(e -> patchedState.putIfAbsent(e.getKey(), e.getValue()));
-        return new Snapshot(patchedState, index);
-    }
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));*/
 
-    void merge(Snapshot snapshot) {
-        state.clear();
-        state.putAll(snapshot.state);
-        index = snapshot.index;
+        return new Snapshot(this, diff, index);
     }
 
     boolean isConsistent() {
-        long stateValue = state.values().stream().reduce(Math::addExact).orElse(Long.MAX_VALUE);
-        if(stateValue != TransactionViewModel.SUPPLY) {
+        if (isConsistentMarker.get()) {
+            return true;
+        }
+        Map<Hash, Long> totalState;
```

Comment:
it seems wasteful that the `totalState` is calculated, but not stored in the snapshot itself.
this means that, for example, `buildSnapshot` recalculates the total state from initial snapshot each for each milestone being rebuilt - O(n^2).